File: library/src/main/java/org/crispy_fish/query/PaxTimeResultsQuery.java
package org.crispy_fish.query;

import org.coner.crispy_fish.domain.Result;
import org.coner.crispy_fish.domain.Driver;
import org.coner.crispy_fish.domain.Numbers;
import org.coner.crispy_fish.domain.Run;
import org.coner.crispy_fish.filetype.ecf.EventControlFile;
import org.crispy_fish.filetype.staging.StagingLineDomainReader;
import org.crispy_fish.filetype.staging.StagingLineReader;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class PaxTimeResultsQuery {

    private final EventControlFile eventControlFile;
    private final StagingLineReader<String> stagingLineReader;
    private final StagingLineDomainReader<String> stagingLineDomainReader;

    public PaxTimeResultsQuery(EventControlFile eventControlFile, StagingLineReader<String> stagingLineReader, StagingLineDomainReader<String> stagingLineDomainReader) {
        this.eventControlFile = eventControlFile;
        this.stagingLineReader = stagingLineReader;
        this.stagingLineDomainReader = stagingLineDomainReader;
    }

    public List<Result> query(List<String> stagingFileLines) throws QueryException {
        Map<Numbers, Result> driverBestPaxResults = new HashMap<>();
        for (String stagingFileLine : stagingFileLines) {
            Driver driver = stagingLineDomainReader.readDriver(stagingFileLine);
            Run run = stagingLineDomainReader.readRun(stagingFileLine);
            if (driver == null || run == null) {
                continue;
            }
            run.setTimeScratchAsString(stagingLineReader.getRunPaxTime(stagingFileLine));
            run.setTimeScratchAsDuration(run.getPaxTime());
            Duration penaltyDuration = Duration.ZERO;
            switch (run.getPenaltyType()) {
                case DID_NOT_FINISH:
                    penaltyDuration = Duration.ofDays(1);
                    break;
                case DISQUALIFIED:
                    penaltyDuration = Duration.ofDays(2);
                    break;
                case RERUN:
                    continue;
                case CLEAN:
                case CONE:
                    // cone penalties already included in pax time
                    break;
                default:
                    throw new IllegalStateException("Unrecognized penalty type: " + run.getPenaltyType());
            }
            if (run.getPaxTime() == null && penaltyDuration != Duration.ZERO) {
                run.setPaxTime(Duration.ZERO);
            }
            run.setTimeScored(run.getPaxTime().plus(penaltyDuration));
            boolean shouldPutResult;
            if (driverBestPaxResults.containsKey(driver.getNumbers())) {
                Result bestResult = driverBestPaxResults.get(driver.getNumbers());
                shouldPutResult = run.getTimeScored().compareTo(bestResult.getRun().getTimeScored()) < 0;
            } else {
                shouldPutResult = true;
            }
            if (!shouldPutResult) {
                continue;
            }
            Result result = new Result();
            result.setDriver(driver);
            result.setRun(run);
            driverBestPaxResults.put(driver.getNumbers(), result);
        }
        List<Result> results = driverBestPaxResults.values().stream()
                .sorted((result1, result2) -> result1.getRun().getTimeScored().compareTo(result2.getRun().getTimeScored()))
                .collect(Collectors.toList());
        int position = 1;
        for (Result result : results) {
            result.setPosition(position++);
        }
        return results;
    }

}

File: library/src/main/kotlin/org/coner/crispy_fish/query/PaxTimeResultsQuery.kt
package org.coner.crispy_fish.query

import org.coner.crispy_fish.domain.*
import org.coner.crispy_fish.filetype.ecf.EventControlFile
import org.crispy_fish.filetype.staging.StagingLineDomainReader
import org.crispy_fish.filetype.staging.StagingLineReader

import java.time.Duration
import java.util.HashMap
import kotlin.streams.toList

class PaxTimeResultsQuery(private val eventControlFile: EventControlFile, private val stagingLineReader: StagingLineReader<String>, private val stagingLineDomainReader: StagingLineDomainReader<String>) {

    @Throws(QueryException::class)
    fun query(stagingFileLines: List<String>): List<Result> {
        val driverBestPaxResults = HashMap<Numbers, Result>()
        for (stagingFileLine in stagingFileLines) {
            val driver = stagingLineDomainReader.readDriver(stagingFileLine)
            val run = stagingLineDomainReader.readRun(stagingFileLine)
            if (driver == null || run == null) {
                continue
            }
            if (run.penaltyType?.shouldOmitRunFromResults == true) continue
            run.timeScratchAsString = stagingLineReader.getRunPaxTime(stagingFileLine)
            run.timeScratchAsDuration = run.paxTime
            var penaltyDuration = Duration.ZERO
            when (run.penaltyType) {
                PenaltyType.CONE -> {
                    // no-op
                    // pax time from staging file already has penalty time added
                }
                PenaltyType.DID_NOT_FINISH -> penaltyDuration = Duration.ofDays(1)
                PenaltyType.DISQUALIFIED -> penaltyDuration = Duration.ofDays(2)
                PenaltyType.RERUN -> throw IllegalStateException("Run should be omitted from results")
                PenaltyType.CLEAN-> { /* no-op */ }
                else -> throw IllegalStateException("Unrecognized penalty type: " + run.penaltyType!!)
            }// cone penalties already included in pax time
            if (run.paxTime == null && penaltyDuration !== Duration.ZERO) {
                run.paxTime = Duration.ZERO
            }
            run.timeScored = run.paxTime!!.plus(penaltyDuration)
            val shouldPutResult: Boolean
            shouldPutResult = if (driverBestPaxResults.containsKey(driver.numbers)) {
                val bestResult = driverBestPaxResults[driver.numbers]
                run.timeScored?.compareTo(bestResult?.run?.timeScored!!) ?: 0 < 0
            } else {
                true
            }
            if (!shouldPutResult) {
                continue
            }
            val result = Result()
            result.driver = driver
            result.run = run
            driverBestPaxResults[driver.numbers!!] = result
        }
        val results = driverBestPaxResults.values.stream()
                .sorted(compareBy { it.run?.timeScored })
                .toList()
        var position = 1
        for (result in results) {
            result.position = position++
        }
        return results
    }

}
